bip-txhash.md

```diff
@@ -0,0 +1,269 @@
+  * the `0x00` byte: it is set equal to `TXFS_SPECIAL_ALL`, which means "ALL" and is primarily
+    useful to emulate `SIGHASH_ALL` when `OP_TXHASH` is used in combination
+    with `OP_CHECKSIGFROMSTACK`.<br>Special case `TXFS_SPECIAL_TEMPLATE` is 4
+    bytes long, as follows:
+    * &nbsp;1. `TXFS_ALL`
+    * &nbsp;2. `TXFS_INPUTS_ALL | TXFS_OUTPUTS_ALL`
+    * &nbsp;3. `TXFS_INOUT_NUMBER | TXFS_INOUT_SELECTION_ALL`
+    * &nbsp;4. `TXFS_INOUT_NUMBER | TXFS_INOUT_SELECTION_ALL`
+
+* The first byte of the TxFieldSelector has its 8 bits assigned as follows, from lowest to highest:
+  * &nbsp;1. version (`TXFS_VERSION`)
+  * &nbsp;2. locktime (`TXFS_LOCKTIME`)
+  * &nbsp;3. current input index (`TXFS_CURRENT_INPUT_IDX`)
+  * &nbsp;4. current input control block (or empty) (`TXFS_CURRENT_INPUT_CONTROL_BLOCK`)
+  * &nbsp;5. current script last `OP_CODESEPARATOR` position (or 0xffffffff)
+    (`TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS`)
+  * &nbsp;6. inputs (`TXFS_INPUTS`)
+  * &nbsp;7. outputs (`TXFS_OUTPUTS`)
+
+* The last (highest) bit of the first byte (`TXFS_CONTROL`), we will call the
+  "control bit", and it can be used to control the behavior of the opcode. For
+  `OP_TXHASH` and `OP_CHECKTXHASHVERIFY`, the control bit is used to determine
+  whether the TxFieldSelector itself has to be included in the resulting hash.
+  (For potential other uses of the TxFieldSelector (like a hypothetical
+  `OP_TX`), this bit can be repurposed.)
+
+* If either "inputs" or "outputs" is set to 1, expect another byte with its 8
+  bits assigning the following variables, from lowest to highest:
+  * Specifying which fields of the inputs will be selected:
+    * &nbsp;1. prevouts (`TXFS_INPUTS_PREVOUTS`)
+    * &nbsp;2. sequences (`TXFS_INPUTS_SEQUENCES`)
+    * &nbsp;3. scriptSigs (`TXFS_INPUTS_SCRIPTSIGS`)
+    * &nbsp;4. prevout scriptPubkeys (`TXFS_INPUTS_PREV_SCRIPTPUBKEYS`)
+    * &nbsp;5. prevout values (`TXFS_INPUTS_PREV_VALUED`)
+    * &nbsp;6. taproot annexes (`TXFS_INPUTS_TAPROOT_ANNEXES`)
+
+  * Specifying which fields of the outputs will be selected:
+    * &nbsp;7. scriptPubkeys (`TXFS_OUTPUTS_SCRIPTPUBKEYS`)
+    * &nbsp;8. values (`TXFS_OUTPUTS_VALUES`)
+
+* We define as follows:
+  * `TXFS_ALL = TXFS_VERSION | TXFS_LOCKTIME | TXFS_CURRENT_INPUT_IDX | TXFS_CURRENT_INPUT_CONTROL_BLOCK | TXFS_CURRENT_INPUT_LAST_CODESEPARATOR_POS | TXFS_INPUTS | TXFS_OUTPUTS | TXFS_CONTROL`
+  * `TXFS_INPUTS_ALL = TXFS_INPUTS_PREVOUTS | TXFS_INPUTS_SEQUENCES | TXFS_INPUTS_SCRIPTSIGS | TXFS_INPUTS_PREV_SCRIPTPUBKEYS | TXFS_INPUTS_PREV_VALUES | TXFS_INPUTS_TAPROOT_ANNEXES`
+  * `TXFS_INPUTS_TEMPLATE = TXFS_INPUTS_SEQUENCES | TXFS_INPUTS_SCRIPTSIGS | TXFS_INPUTS_PREV_VALUES | TXFS_INPUTS_TAPROOT_ANNEXES`
+  * `TXFS_OUTPUTS_ALL = TXFS_OUTPUTS_SCRIPTPUBKEYS | TXFS_OUTPUTS_VALUES`
+
+For both inputs and then outputs, do the following:
+
+* If the "in/outputs" field is set to 1, another additional byte is expected:
```

Comment:
If I'm following this correctly, the (non-special) TxFieldSelector format is, in bytes:

CORE_SELECTOR [INOUT_SELECTOR] [IN_SELECTOR] [OUT_SELECTOR]

If `TXFS_INPUTS` is set in the CORE_SELECTOR, then INOUT_SELECTOR and IN_SELECTOR are present.  If `TXFS_OUTPUTS` is set in CORE_SELECTOR, then INOUT_SELECTOR and OUT_SELECTOR are present?

Reply:
Exactly. I'm thinking of changing this as follows:

- remove `TXFS_INPUTS` and `TXFS_OUTPUTS` bits
- reader will know the entire size of the txfs, so when a second byte is present, look at the bits present in the `INOUT_SELECTOR` byte to know whether to expect `IN_SELECTOR` and/or `OUT_SELECTOR`.
- this frees up two bits in the `CORE_SELECTOR`, one of which I'm thinking to repurpose for `SPEND_SCRIPT` (i.e. scriptCode for segwit v0 inputs and tapscript for v1 inputs, scriptPubkey for non-segwit)